gitolize.sh

```diff
@@ -0,0 +1,167 @@
+#!/bin/bash
+
+# A script wrapping a shell command so that it runs with a directory populated from a git repository.
+# In the write mode, the repository is locked with a temporary file and
+# changes made in that directory by the shell command get pushed back to the repository.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+usage() {
+  echo "Usage: $0 [-b <branch>] [-d <git_directory>] [-l <local_directory>] [-m message] [-w] [command...]" 1>&2
+  exit 1
+}
+
+git() {
+  COMMAND="$1"
+  shift
+  EXTRA_FLAGS=
+  case "$COMMAND" in
+    add)
+      EXTRA_FLAGS="$GIT_VERBOSE_FLAG"
+      ;;
+    *)
+      EXTRA_FLAGS="$GIT_QUIET_FLAG"
+      ;;
+  esac
+  if [[ "$VERBOSE" ]]
+  then
+    echo "running $GIT -C "$LOCAL_DIRECTORY" $COMMAND $EXTRA_FLAGS $@" 1>&2
+  fi
+  $GIT -C "$LOCAL_DIRECTORY" $COMMAND $EXTRA_FLAGS "$@"
+}
+
+LOCK_FILE=gitolize.lock
+
+EXIT_CODE=0
+
+GIT="$(which git)"
+GIT_QUIET_FLAG=--quiet
+GIT_VERBOSE_FLAG=
+
+BRANCH=main
+GIT_DIRECTORY=.
+LOCAL_DIRECTORY=
+CLEANUP_LOCAL_DIRECTORY=
+COMMIT_MESSAGE=
+VERBOSE=
+WRITE=
+
+
+while getopts b:d:l:m:vw OPTION
+do
+  case $OPTION in
+    b)
+      BRANCH="$OPTARG"
+      ;;
+    d)
+      GIT_DIRECTORY="$OPTARG"
+      ;;
+    l)
+      LOCAL_DIRECTORY="$OPTARG"
+      ;;
+    m)
+      COMMIT_MESSAGE="$OPTARG"
+      ;;
+    v)
+      VERBOSE=true
+      ;;
+    w)
+      WRITE=true
+      ;;
+    *)
+      usage
+      ;;
+  esac
+done
+
+shift $((OPTIND-1))
+
+if [[ "$#" < 2 ]]
+then
+  usage
+fi
+
+if [[ ! "$LOCAL_DIRECTORY" ]]
+then
+  LOCAL_DIRECTORY="$(mktemp -d)"
+  CLEANUP_LOCAL_DIRECTORY=true
+fi
+
+# /. is illegal, see https://git-scm.com/docs/git-check-ref-format#_description
+LOCK_BRANCH_SUFFIX="${GIT_DIRECTORY/./}"
+if [[ "$LOCK_BRANCH_SUFFIX" ]]
+then
+   LOCK_BRANCH="lock/$LOCK_BRANCH_SUFFIX"
+else
+   LOCK_BRANCH="lock"
+fi
+
+if [[ "$VERBOSE" ]]
+then
+  GIT_QUIET_FLAG=
+  GIT_VERBOSE_FLAG=--verbose
+fi
+
+GIT_REPOSITORY="$1"
+shift
+
+if [[ ! "$COMMIT_MESSAGE" ]]
```

Comment:
We could extend the commit message to include:
- `$GIT_DIRECTORY`
- run details like elapsed time

Reply:
In an offline chat you suggested also adding the stdin & stdout of the command to the commit message. It would add a lot of useful context information with the caveat of sometimes being really large (I'm wondering if there is a limit for the commit message length) and including sensitive information. Maybe it would be enabled with an explicit flag?